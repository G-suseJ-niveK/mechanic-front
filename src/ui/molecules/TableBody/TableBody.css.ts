import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
  bodyCell: {
    padding: '10px',
    with: '200px'
  },
  textNoItems: {
    fontSize: '14px',
    padding: '15px',
    fontWeight: 'bold'
    // borderBottom: '0.14em solid #000000'
  },
  celda: {
    borderBottom: '1px solid #0000001d'
  }
});

export default useStyles;
