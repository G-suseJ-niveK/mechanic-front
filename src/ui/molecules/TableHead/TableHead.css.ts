import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
  rootCell: {
    color: '#446125',
    padding: '10px 10px',
    borderBottom: '2px #red solid'
  },
  celda: {
    borderBottom: '2px solid #c9c8c8'
  }
});

export default useStyles;
