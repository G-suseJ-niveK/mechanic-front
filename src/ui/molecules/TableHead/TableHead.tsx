import React from 'react';
import { TableRow, TableCell, TableSortLabel, TableHead } from '@material-ui/core';
import useStyles from './TableHead.css';

/**
 * ITableHeadColumn define los params de cada columna de la tabla
 */
export type TableHeadColumn = {
  sorteable: boolean;
  align?: 'left' | 'right' | 'inherit' | 'center' | 'justify' | undefined;
  padding?: 'checkbox' | 'none' | 'normal' | undefined;
  size?: 'medium' | 'small' | undefined;
  variant?: 'body' | 'footer' | 'head' | undefined;
  text: string;
  value: string;
  render?: (value: any) => void;
};

/**
 * ITableHeadProps
 */
export type TableHeadProps = {
  headers: TableHeadColumn[];
  orderBy: string;
  order: 'asc' | 'desc' | undefined;
  createSortHandler: (column: string) => void;
};

const CustomTableHead: React.FC<TableHeadProps> = (props: TableHeadProps) => {
  const classes = useStyles();
  const { headers, orderBy, order, createSortHandler }: TableHeadProps = props;
  function handleSort(e: React.FormEvent, column: string) {
    createSortHandler(column);
  }

  return (
    <>
      <TableHead>
        <TableRow>
          {headers.map((column: TableHeadColumn, index: number) => {
            const { sorteable, align, padding, size, variant, text, value } = column;
            return (
              <TableCell
                className={classes.rootCell}
                key={`table_column_${index}`}
                align={align}
                padding={padding}
                sortDirection={orderBy === value ? order : false} //esta bien?
                size={size}
                variant={variant}>
                {sorteable ? (
                  <TableSortLabel
                    active={orderBy === value}
                    direction={order}
                    onClick={(e: any) => handleSort(e, value)}>
                    {text}
                  </TableSortLabel>
                ) : (
                  <>{text}</>
                )}
              </TableCell>
            );
          })}
        </TableRow>
      </TableHead>
    </>
  );
};
CustomTableHead.defaultProps = {
  createSortHandler: () => null,
  headers: [],
  order: undefined,
  orderBy: ''
};
export default React.memo(CustomTableHead);
