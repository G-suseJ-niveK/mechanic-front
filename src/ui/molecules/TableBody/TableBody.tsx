import React from 'react';
import { TableHeadColumn } from '../TableHead/TableHead';
import { TableCell, TableRow, TableBody, CircularProgress } from '@material-ui/core';
import useStyles from './TableBody.css';
//import { render } from '@testing-library/react'

/**
 * TableBodyActions define los atributos para las acciones de cada botón en la tabla (en este caso botones)
 */
export type TableBodyActions = {
  onClick: (value: any, index: number) => void;
  text: string;
  tooltip?: string;
  icon: React.ReactElement;
};

/**
 * se obtienen los (headers) de la tabla, teniendo en cuenta el value "actions" el cual define si la tabla
 * posee acciones
 * param actions[] define los botones que posee cada fila
 */
export type TableBodyProps = {
  headers: TableHeadColumn[];
  items: any[];
  textNoItems?: string;
  loading?: boolean;
};

const CustomTableBody: React.FC<TableBodyProps> = (props: TableBodyProps) => {
  const classes = useStyles();
  const { items, headers, textNoItems, loading }: TableBodyProps = props;
  return (
    <>
      <TableBody>
        {loading ? (
          <TableRow className={classes.celda}>
            <TableCell align="center" colSpan={headers.length}>
              <CircularProgress color="primary" />
            </TableCell>
          </TableRow>
        ) : items.length > 0 ? (
          items.map((row: any, rowIndex: number) => {
            return (
              <TableRow key={`table_body_row_${rowIndex}`} hover={true} className={classes.celda}>
                {headers.map((column: any, columnIndex: number) => {
                  /* evalúa si el head es de acciones no se renderiza aquí, sino en el siguiente */

                  return (
                    <TableCell align={column.align} key={`table_value_${columnIndex}`} className={classes.bodyCell}>
                      {/* valida si render esta disponible */}
                      {column.render ? column.render(row) : row[column.value]}
                    </TableCell>
                  );
                })}
              </TableRow>
            );
          })
        ) : (
          <TableRow>
            <TableCell align="center" colSpan={headers.length} className={classes.textNoItems}>
              {textNoItems}
            </TableCell>
          </TableRow>
        )}
      </TableBody>
    </>
  );
};

CustomTableBody.defaultProps = {
  headers: [],
  items: [],
  loading: false,
  textNoItems: 'No se encontraron resultados.'
};

export default React.memo(CustomTableBody);
