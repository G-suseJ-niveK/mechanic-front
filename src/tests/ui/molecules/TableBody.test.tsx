import React from 'react';
import { render, screen } from '@testing-library/react';
import TableBody from '~molecules/TableBody/TableBody';
import { TableHeadColumn } from '~molecules/TableHead/TableHead';

describe('Render TableBody', () => {
  // const textNoItems: string = 'mock-text-no-items';
  const headers: TableHeadColumn[] = [
    {
      sorteable: true,
      align: 'left',
      padding: 'none',
      size: 'medium',
      variant: 'body',
      text: 'description',
      value: 'description'
    }
  ];

  const items = [
    {
      id: 1,
      description: 'mock-description-1'
    },
    {
      id: 2,
      description: 'mock-description-2'
    }
  ];

  test('should render', () => {
    const { container } = render(
      <table>
        <TableBody headers={[]} items={[]} />
      </table>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render headers and items', () => {
    render(
      <table>
        <TableBody headers={headers} items={items} />
      </table>
    );

    expect(screen.getByRole('row', { name: 'mock-description-1' })).toBeInstanceOf(Node);
    expect(screen.getByRole('row', { name: 'mock-description-2' })).toBeInstanceOf(Node);
    expect(screen.getByRole('cell', { name: 'mock-description-1' })).toBeInstanceOf(Node);
    expect(screen.getByRole('cell', { name: 'mock-description-2' })).toBeInstanceOf(Node);
  });
});
