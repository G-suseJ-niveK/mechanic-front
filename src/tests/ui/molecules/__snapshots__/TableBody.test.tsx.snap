// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render TableBody should render 1`] = `
<div>
  <table>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-textNoItems-2 MuiTableCell-alignCenter"
          colspan="0"
        >
          No se encontraron resultados.
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
